extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -0,0 +1,259 @@
+            case TIME:
+                return io.deephaven.qst.type.Type.find(java.time.LocalTime.class);
+            case DECIMAL:
+                return io.deephaven.qst.type.Type.find(java.math.BigDecimal.class);
+            case FIXED: // Fall through
+            case BINARY:
+                return io.deephaven.qst.type.Type.find(byte[].class);
+            default:
+                throw new TableDataException("Unsupported iceberg column type " + typeId.name());
+        }
+    }
+
+    @SuppressWarnings("unused")
+    public List<TableIdentifier> listTables(final Namespace namespace) {
+        // TODO: have this return a Deephaven Table of table identifiers
+        return catalog.listTables(namespace);
+    }
+
+    public List<Long> listTableSnapshots(@NotNull final TableIdentifier tableIdentifier) {
+        final ArrayList<Long> snapshotIds = new ArrayList<>();
+        catalog.loadTable(tableIdentifier).snapshots().forEach(snapshot -> snapshotIds.add(snapshot.snapshotId()));
+        return snapshotIds;
+    }
+
+    /**
+     * Read the latest static snapshot of a table from the Iceberg catalog.
+     *
+     * @param tableIdentifier The table identifier to load
+     * @return The loaded table
+     */
+    @SuppressWarnings("unused")
+    public Table snapshotTable(@NotNull final TableIdentifier tableIdentifier) {
+        return readTableInternal(tableIdentifier, -1, false);
+    }
+
+    /**
+     * Read a static snapshot of a table from the Iceberg catalog.
+     *
+     * @param tableIdentifier The table identifier to load
+     * @param snapshotId The snapshot ID to load
+     * @return The loaded table
+     */
+    @SuppressWarnings("unused")
+    public Table snapshotTable(
+            @NotNull final TableIdentifier tableIdentifier,
+            final long snapshotId) {
+        return readTableInternal(tableIdentifier, snapshotId, false);
+    }
+
+    private Table readTableInternal(
+            @NotNull final TableIdentifier tableIdentifier,
+            final long snapshotId,
+            final boolean isRefreshing) {
+
+        // Load the table from the catalog
+        final org.apache.iceberg.Table table = catalog.loadTable(tableIdentifier);
+
+        // Do we want the latest or a specific snapshot?
+        final Snapshot snapshot;
+        if (snapshotId < 0) {
+            snapshot = table.currentSnapshot();
+        } else {
+            snapshot = table.snapshot(snapshotId);
+        }
+        final Schema schema = table.schemas().get(snapshot.schemaId());
+
+        // Load the partitioning schema
+        final org.apache.iceberg.PartitionSpec partitionSpec = table.spec();
+
+        // Use the user-supplied defininition or the Iceberg schema for the table definition.
+        final TableDefinition tableDef = instructions.tableDefinition().isPresent()
+                ? instructions.tableDefinition().get()
+                : fromSchema(schema, partitionSpec, instructions);
```

Comment:
We should probably make the definition regardless and do some kind of validity check. We only care about the intersection of the columns (by name). `TableDefinition` has methods, but not this exact one.

You could do a `fromSchema` that filters columns not in the `TableDefinition's` name map, which would let you defer to `checkCompatibility()`.